#!/usr/local/bin/Rscript
# Fetched from https://github.com/PapenfussLab/gridss/blob/master/example/simple-event-annotation.R
# Although it's been slightly adjusted

# if (!requireNamespace("BiocManager", quietly = TRUE)) install.packages("BiocManager")
#BiocManager::install("StructuralVariantAnnotation")
#install.packages("stringr")
library(VariantAnnotation)
library(StructuralVariantAnnotation)
library(stringr)

args <- commandArgs()
input_vcf <- args[1]
output_vcf <- args[2]

#' Simple SV type classifier
simpleEventType <- function(gr) {
    pgr = partner(gr)
    return(ifelse(seqnames(gr) != seqnames(pgr), "CTX", # inter-chromosomosal
        ifelse(strand(gr) == strand(pgr), "INV",
        ifelse(gr$insLen >= abs(gr$svLen) * 0.7, "INS", # TODO: improve classification of complex events
        ifelse(xor(start(gr) < start(pgr), strand(gr) == "-"), "DEL",
        "DUP")))))
}

vcf <- readVcf(input_vcf, "hg38")
info(header(vcf)) = unique(as(rbind(as.data.frame(info(header(vcf))), data.frame(
    row.names=c("SIMPLE_TYPE"),
    Number=c("1"),
    Type=c("String"),
    Description=c("Simple event type annotation based purely on breakend position and orientation."))), "DataFrame"))
gr <- breakpointRanges(vcf)
svtype <- simpleEventType(gr)
info(vcf)$SIMPLE_TYPE <- NA_character_
info(vcf[gr$sourceId])$SIMPLE_TYPE <- svtype
info(vcf[gr$sourceId])$SVLEN <- gr$svLen
writeVcf(vcf, output_vcf) # generated by example/gridss.sh

# # TODO: perform event filtering here
# # By default, GRIDSS is very sensitive but this comes at the cost of a high false discovery rate
# gr <- gr[gr$FILTER == "PASS" & partner(gr)$FILTER == "PASS"] # Remove low confidence calls

# simplegr <- gr[simpleEventType(gr) %in% c("INS", "INV", "DEL", "DUP")]
# simplebed <- data.frame(
#     chrom=seqnames(simplegr),
#     # call the centre of the homology/inexact interval
#     start=as.integer((start(simplegr) + end(simplegr)) / 2),
#     end=as.integer((start(partner(simplegr)) + end(partner(simplegr))) / 2),
#     name=simpleEventType(simplegr),
#     score=simplegr$QUAL,
#     strand="."
# )
# # Just the lower of the two breakends so we don't output everything twice
# simplebed <- simplebed[simplebed$start < simplebed$end,]
# write.table(simplebed, "chr12.1527326.DEL1024.simple.bed", quote=FALSE, sep='\t', row.names=FALSE, col.names=FALSE)
