Changes in module 'nf-core/gawk'
--- modules/nf-core/gawk/main.nf
+++ modules/nf-core/gawk/main.nf
@@ -22,7 +22,7 @@
     def args  = task.ext.args  ?: '' // args is used for the main arguments of the tool
     def args2 = task.ext.args2 ?: '' // args2 is used to specify a program when no program file has been given
     prefix = task.ext.prefix ?: "${meta.id}"
-    suffix = task.ext.suffix ?: "${input.getExtension}"
+    suffix = task.ext.suffix ?: "${input.getExtension()}"
 
     program = program_file ? "-f ${program_file}" : "${args2}"
 

************************************************************
