Changes in module 'nf-core/annotsv/annotsv'
--- modules/nf-core/annotsv/annotsv/main.nf
+++ modules/nf-core/annotsv/annotsv/main.nf
@@ -45,6 +45,14 @@
 
     mv *_AnnotSV/* .
 
+    if [ ! -f "${prefix}.tsv ]; then
+        touch ${prefix}.tsv
+    fi
+
+    if [ ! -f "${prefix}.vcf ]; then
+        touch ${prefix}.vcf
+    fi
+
     cat <<-END_VERSIONS > versions.yml
     "${task.process}":
         annotsv: \$(echo \$(AnnotSV -help 2>&1 | head -n1 | sed 's/^AnnotSV //'))

************************************************************
