Changes in module 'nf-core/samtools/merge'
--- modules/nf-core/samtools/merge/main.nf
+++ modules/nf-core/samtools/merge/main.nf
@@ -13,10 +13,9 @@
     tuple val(meta3), path(fai)
 
     output:
-    tuple val(meta), path("${prefix}.bam") , optional:true, emit: bam
-    tuple val(meta), path("${prefix}.cram"), optional:true, emit: cram
-    tuple val(meta), path("*.csi")         , optional:true, emit: csi
-    path  "versions.yml"                                  , emit: versions
+    tuple val(meta), path("*.{bam,cram}") , optional:true, emit: bam
+    tuple val(meta), path("*.csi")        , optional:true, emit: csi
+    path  "versions.yml"                  , emit: versions
 
 
     when:

************************************************************
