Changes in module 'nf-core/bcftools/filter'
--- modules/nf-core/bcftools/filter/main.nf
+++ modules/nf-core/bcftools/filter/main.nf
@@ -43,6 +43,7 @@
     """
 
     stub:
+    def args = task.ext.args ?: ''
     def prefix = task.ext.prefix ?: "${meta.id}"
 
     extension = args.contains("--output-type b") || args.contains("-Ob") ? "bcf.gz" :

************************************************************
