Changes in module 'nf-core/bcftools/concat'
--- modules/nf-core/bcftools/concat/main.nf
+++ modules/nf-core/bcftools/concat/main.nf
@@ -8,7 +8,7 @@
         'biocontainers/bcftools:1.17--haef29d1_0' }"
 
     input:
-    tuple val(meta), path(vcfs), path(tbi)
+    tuple val(meta), path(vcfs), path(tbis)
 
     output:
     tuple val(meta), path("*.gz"), emit: vcf
@@ -19,8 +19,19 @@
 
     script:
     def args = task.ext.args   ?: ''
-    prefix   = task.ext.prefix ?: "${meta.id}"
+    def prefix   = task.ext.prefix ?: "${meta.id}"
+
+    def tbi_names = tbis.collect { it.name }
+
+    def tabix_vcfs = vcfs.collect {
+        if(tbi_names.contains("${it.name}.tbi" as String)) {
+            return ""
+        }
+        return "tabix ${it.name}"
+    }
+
     """
+    ${tabix_vcfs.join("\n")}
     bcftools concat \\
         --output ${prefix}.vcf.gz \\
         $args \\

************************************************************
