Changes in module 'nf-core/bcftools/concat'
--- modules/nf-core/bcftools/concat/main.nf
+++ modules/nf-core/bcftools/concat/main.nf
@@ -2,13 +2,13 @@
     tag "$meta.id"
     label 'process_medium'
 
-    conda "bioconda::bcftools=1.17"
+    conda "${moduleDir}/environment.yml"
     container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
-        'https://depot.galaxyproject.org/singularity/bcftools:1.17--haef29d1_0':
-        'biocontainers/bcftools:1.17--haef29d1_0' }"
+        'https://depot.galaxyproject.org/singularity/bcftools:1.18--h8b25389_0':                                                                                                                                                            
+        'biocontainers/bcftools:1.18--h8b25389_0' }" 
 
     input:
-    tuple val(meta), path(vcfs), path(tbi)
+    tuple val(meta), path(vcfs), path(tbis)
 
     output:
     tuple val(meta), path("*.gz"), emit: vcf
@@ -19,13 +19,28 @@
 
     script:
     def args = task.ext.args   ?: ''
-    prefix   = task.ext.prefix ?: "${meta.id}"
+    def args2 = task.ext.args2   ?: ''
+    def prefix   = task.ext.prefix ?: "${meta.id}"
+
+    def tbi_names = tbis.collect { it.name }
+
+    def tabix_vcfs = vcfs.collect {
+        if(tbi_names.contains("${it.name}.tbi" as String)) {
+            return ""
+        }
+        return "tabix ${it.name}"
+    }
+
     """
+    ${tabix_vcfs.join("\n")}
     bcftools concat \\
-        --output ${prefix}.vcf.gz \\
         $args \\
         --threads $task.cpus \\
-        ${vcfs}
+        ${vcfs} \\
+    | bcftools sort \\
+        ${args2} \\
+        --output ${prefix}.vcf.gz \\
+        --output-type z
 
     cat <<-END_VERSIONS > versions.yml
     "${task.process}":

--- /dev/null
+++ modules/nf-core/bcftools/concat/environment.yml
@@ -0,0 +1,7 @@
+name: bcftools_concat
+channels:
+  - conda-forge
+  - bioconda
+  - defaults
+dependencies:
+  - bioconda::bcftools=1.18

************************************************************
